Asymetric keys as Core Entities

javierguerragiraldez opened the pull request.

mikefero commented:
@javierguerragiraldez Did a very high level review of this PR and excellent job. I think we should add integration tests for validating the key/key-sets against a 3.1.0 data-plane and ensure that version compatibility is properly handled for < 3.1.0 data-planes.

GGabriele commented on internal/grpc/proto/kong/admin/model/v1/key.proto:
```diff
@@ -0,0 +1,65 @@
+  string name = 6;
+  JwkKey jwk = 7;
+  PemKey pem = 8;
+  KeyFormat key_format = 9;
```
Is this field used at all?

javierguerragiraldez replied:
I don't think so; but we'll have to check with the Kong implementation (which i think has just been merged in Kong/kong#9737)

GGabriele commented on internal/model/json/validation/typedefs/fields.go:
```diff
@@ -217,4 +217,44 @@ var Reference = &generator.Schema{
+var PemKey = &generator.Schema{
+	Type: "object",
+	Properties: map[string]*generator.Schema{
+		"private_key": {Type: "string"},
```
You can use [some existing formats ](https://github.com/Kong/koko/blob/main/internal/model/json/schema/formats.go#L71-L73)for these

Example: https://github.com/Kong/koko/blob/main/internal/resource/certificate.go#L180

GGabriele commented on internal/server/kong/ws/config/compat/compatibility_table.go:
```diff
@@ -869,6 +869,36 @@ var (
+				Name:         "keys",
+				Type:         config.TopLevel,
+				RemoveFields: []string{"keys"},
+				Remove:       true,
```
We should test this extensively, but at a first glance this is mixing 2 concept we have been using in 2 different ways: `RemoveFields` is used to remove some entity fields. while `Remove` is used to remove the whole entity. I guess that the latter is useless after your introduction of `TopLevel`?

javierguerragiraldez replied:
i couldn't find a way to remove a whole "thing", so i added that `TopLevel`, which is a different code, so it does reuse other fields for a slightly different purpose.  in this case, those are the fields of the top level to remove.  the `Remove` field is unused right now.

(would it be too bad taste to say that all this makes me miss XSLT?)

mikefero replied:
Could we not [utilize the current implementation that exists](https://github.com/Kong/koko/blob/f761e0bd4ba0437f40c5e6504c62065adc496f75/internal/server/kong/ws/config/version_compatibility.go#L32-L52) and define `Keys` and `KeySets` instead of `TopLevel`?

javierguerragiraldez replied:
didn't realize the `Remove` flag actually removed the whole type of entities.

done

GGabriele reviewed (commented):
A few high level notes:
- so far we have been sticking to one-file (proto, resouce, service, tests etc) per entity. Do you think would it be possible to split `key` and `keyset` into two separate files (again for proto, resouce, service, tests etc)?
- we should test `ValidateKeySchema` and `ValidateKeySetSchema` calls too
- can you please add more test cases using more key fields?

Other than that, great job!

javierguerragiraldez commented:
> so far we have been sticking to one-file (proto, resouce, service, tests etc) per entity. Do you think would it be possible to split key and keyset into two separate files (again for proto, resouce, service, tests etc)?

no prob.  just mixed them because key-sets don't have much meaning on their own.

> we should test ValidateKeySchema and ValidateKeySetSchema calls can you please add more test cases using more key fields?

right now there's not much to test there; so any tests would have to come after settling on how would the schemas look.  having the PEM-format checking is a step in the right direction; now i'm fighting with the exclusion (only need one of the `jwt` and `pem` fields, but each should be thoroughly validated if present)

javierguerragiraldez commented:
some new things after the initial "ready for review" yesterday:
- [x] removed the `TopLevel` update type.  just defined new `Key` and `KeySet` types and used the `Remove` flag.
- [x] use the existing schema formats to validate the public and private keys in the PEM field.
- [x] added "JWK or PEM required but not both" rules to the schema.
- [x] split keyset definitions (at all layers) into their own files.

Still missing:
- [ ] any validation on JWK format
- [ ] integration tests with Kong 3.1 to verify it's getting (and accepting) the keys

Also not sure about the myriad of fields in the JWK definition. Are they contained in the canonical encoding? if so, should they all be individually exposed to the API or should we accept encoded keys?

GGabriele commented:
> Also not sure about the myriad of fields in the JWK definition. Are they contained in the canonical encoding? if so, should they all be individually exposed to the API or should we accept encoded keys?

What does Kong do?

GGabriele commented on internal/resource/keyset.go:
```diff
@@ -0,0 +1,100 @@
+
+	keysetSchema := &generator.Schema{
+		Type: "object",
+		Properties: map[string]*generator.Schema{
```
Missing [tags](https://github.com/Kong/kong/pull/9737/files#diff-068404582b1b262395dc45c10cfb91cc7b0ca1dd49b1c75f1baa1618dbaa69beR31)

javierguerragiraldez replied:
:man_facepalming: 
(added)

GGabriele commented on internal/resource/key_test.go:
```diff
@@ -0,0 +1,203 @@
+			},
+		},
+		{
+			name: "key with valid pem is valid",
```
Mind adding a negative test too? e.g. invalid pem is rejected

GGabriele commented on internal/server/kong/ws/config/version_compatibility_test.go:
```diff
@@ -6274,6 +6274,95 @@ func TestVersionCompatibility_ProcessConfigTableUpdates(t *testing.T) {
 			},
 		},
+		{
+			name: "removing whole top-level fields",
```
Can you please add a couple of more tests:
- test that `KeySet` is removed when `< 3.1.0`
- test that `KeySet` is not removed when `>= 3.1.0`
- test that `Key` is removed when `>= 3.1.0`

javierguerragiraldez replied:
I added that test for the `TopLevel` operation type I wrote; but it was noted that that was redundant with the `Remove: true` flag, so this test is actually redundant with the "remove core entities when indicated", and the others you mention are equivalent to the "don't remove core entities when not indicated"

GGabriele commented on internal/server/kong/ws/config/compat/compatibility_table.go:
```diff
@@ -869,6 +869,34 @@ var (
+			},
+			SemverRange: versionsPre310,
+			Update: config.ConfigTableUpdates{
+				Name:   "keys",
```
```suggestion
				Name:   config.Key.String(),
```

javierguerragiraldez replied:
:heavy_check_mark:

GGabriele commented on internal/server/kong/ws/config/compat/compatibility_table.go:
```diff
@@ -869,6 +869,34 @@ var (
+			},
+			SemverRange: versionsPre310,
+			Update: config.ConfigTableUpdates{
+				Name:   "key_sets",
```
```suggestion
				Name:   config.KeySet.String(),
```

javierguerragiraldez replied:
:ballot_box_with_check:

GGabriele commented on internal/test/e2e/version_compatibility_test.go:
```diff
@@ -1283,3 +1283,56 @@ func TestVersionCompatibility_Zipkin310OrNewer(t *testing.T) {
+		kongClient.RunWhenKong(t, "< 3.0.0")
+
+		util.WaitFunc(t, func() error {
+			return util.EnsureConfig(&v1.TestingConfig{})
```
In order to make sure the configuration can be sync'd down, can you please add another entity (like, a `service`?) in the configuration and assert that it shows up in the DP?

GGabriele commented on internal/test/e2e/version_compatibility_test.go:
```diff
@@ -1283,3 +1283,56 @@ func TestVersionCompatibility_Zipkin310OrNewer(t *testing.T) {
+	})
+
+	t.Run("pre 3.1", func(t *testing.T) {
+		kongClient.RunWhenKong(t, ">= 3.0.0 < 3.1.0")
```
Maybe I'm missing something, but couldn't this test and the `pre 3.0` one be converged in a single `pre 3.1`? How is `pre 3.0` different than `pre 3.1`?

javierguerragiraldez replied:
yes, `pre 3.0` is redundant; i added it to see how the current "master" image identifies as.  it seems it still doesn't say 3.1, so the test isn't too useful yet (unless there's some 3.1 image somewhere).

tjasko commented on internal/resource/keyset.go:
```diff
@@ -0,0 +1,101 @@
+
+func (ks KeySet) Indexes() []model.Index {
+	res := []model.Index{
+		{
+			Name:      "id",
+			Type:      model.IndexUnique,
+			Value:     ks.KeySet.Id,
+			FieldName: "id",
+		},
```
Don't need to add this, it's done automatically by the persistence layer.

tjasko commented on internal/resource/key.go:
```diff
@@ -0,0 +1,133 @@
+
+func (k Key) Indexes() []model.Index {
+	res := []model.Index{
+		{
+			Name:      "id",
+			Type:      model.IndexUnique,
+			Value:     k.Key.Id,
+			FieldName: "id",
+		},
```
Don't need to add this, it's done automatically by the persistence layer.

tjasko commented on internal/resource/key.go:
```diff
@@ -0,0 +1,133 @@
+			"tags":       typedefs.Tags,
+		},
+		AdditionalProperties: &falsy,
+		Required:             []string{"id"},
```
Should mark `name` as required as it's included in with the index. If we don't want it to be required, alter the `Indexes()` method to optionally include it.

tjasko commented on internal/resource/key_test.go:
```diff
@@ -0,0 +1,246 @@
+			k := tt.Key()
+			err := k.Validate(context.Background())
+			if (err != nil) != tt.wantErr {
+				t.Errorf("Validate() error = %v, wantErr %v", err, tt.wantErr)
```
nit: The mixed usages of `assert` and stdlib testing utils is strange, I generally try to stick to one or the other.

tjasko commented on internal/resource/key_test.go:
```diff
@@ -0,0 +1,246 @@
+				t.Errorf("Validate() error = %v, wantErr %v", err, tt.wantErr)
+			}
+			if tt.Errs != nil {
+				verr, _ := err.(validation.Error)
```
There is `require.IsType(t, validation.Error{}, err)` btw.

tjasko commented on internal/server/admin/key_service.go:
```diff
@@ -0,0 +1,144 @@
+	res := resource.NewKey()
+	res.Key = req.Item
+	if err := db.Create(ctx, res); err != nil {
+		s.logger(ctx).Error("error creating", zap.Error(err))
```
Can we have this be a full sentence? e.g.:
```suggestion
		s.logger(ctx).Error("unable to create new key entity", zap.Error(err))
```

tjasko commented on internal/server/admin/key_service.go:
```diff
@@ -0,0 +1,144 @@
+	}, nil
+}
+
+func (s *KeyService) CreateKey(
```
No validation of the ID if they're providing one?

tjasko commented on internal/server/admin/key_test.go:
```diff
@@ -0,0 +1,228 @@
+
+func validateGoodKey(body *httpexpect.Object) {
+	body.ContainsKey("id")
+	body.ContainsKey("created_at")
+	body.ContainsKey("updated_at")
```
nit: Can at least check if it's greater than zero.

tjasko commented on internal/server/admin/key_test.go:
```diff
@@ -0,0 +1,228 @@
+}
+
+func validateGoodKey(body *httpexpect.Object) {
+	body.ContainsKey("id")
```
nit: Can check if it's a valid UUID.

tjasko commented on internal/server/admin/key_test.go:
```diff
@@ -0,0 +1,228 @@
+	t.Run("re-upserting the same key with different id fails",
+		func(_ *testing.T) {
+			k := goodKey()
+			c.PUT("/v1/keys/" + k.Id).WithJSON(k).Expect().Status(http.StatusOK)
```
Something I recently learned about `httpexpect`, it supports placeholders:
```suggestion
			c.PUT("/v1/keys/{id}", k.Id).WithJSON(k).Expect().Status(http.StatusOK)
```

javierguerragiraldez replied:
:+1:

tjasko commented on internal/server/admin/key_test.go:
```diff
@@ -0,0 +1,228 @@
+		func(_ *testing.T) {
+			k := goodKey()
+			c.PUT("/v1/keys/" + k.Id).WithJSON(k).Expect().Status(http.StatusOK)
+			c.PUT("/v1/keys/" + uuid.NewString()).WithJSON(k).Expect().Status(http.StatusBadRequest)
```
Interesting, that really should be HTTP 412... probably not handling the error right, we're blatantly labeling a bunch of errors with HTTP 400 due to our lacking error handling.

tjasko commented on internal/server/admin/keyset_service.go:
```diff
@@ -0,0 +1,145 @@
+	CommonOpts
+}
+
+func (s *KeySetService) GetKeySet(
```
No ID validation?

tjasko commented on internal/server/admin/keyset_service.go:
```diff
@@ -0,0 +1,145 @@
+	}, nil
+}
+
+func (s *KeySetService) CreateKeySet(
```
No ID validation if provided?

GGabriele replied:
Not needed, that's a POST

tjasko commented on internal/server/admin/keyset_service.go:
```diff
@@ -0,0 +1,145 @@
+	res.KeySet = req.Item
+	s.logger(ctx).Debug("copied keyset", zap.Any("keyset", res.KeySet))
+	if err := db.Create(ctx, res); err != nil {
+		s.logger(ctx).Error("error creating", zap.Error(err))
```
Can we have this be a full sentence? e.g.:

```suggestion
		s.logger(ctx).Error("unable to create keyset entity", zap.Error(err))
```

tjasko commented on internal/server/admin/keyset_service.go:
```diff
@@ -0,0 +1,145 @@
+
+	res := resource.NewKeySet()
+	res.KeySet = req.Item
+	s.logger(ctx).Debug("copied keyset", zap.Any("keyset", res.KeySet))
```
What's the use of this debug log message?

tjasko commented on internal/server/kong/ws/config/key_test.go:
```diff
@@ -0,0 +1,80 @@
+
+	opts := MutatorOpts{ClusterID: "xx"}
+
+	err := l.Mutate(context.Background(), opts, config)
```
Should ensure that pagination works.

tjasko commented on internal/server/kong/ws/config/version_compatibility_test.go:
```diff
@@ -6274,6 +6274,95 @@ func TestVersionCompatibility_ProcessConfigTableUpdates(t *testing.T) {
+						Name:     "keys",
+						Type:     Key,
+						Remove:   true,
+						ChangeID: "T102",
```
I don't see any change metadata registered for `T102`?

javierguerragiraldez replied:
these rules are defined for test only; they don't require user-visible metadata.

tjasko replied:
Derp, completely missed over that. Mind me.

tjasko commented on internal/server/kong/ws/config/keyset.go:
```diff
@@ -0,0 +1,57 @@
+	var pageNum int32 = 1
+	var allKeySets []*v1.KeySet
+	for {
+		keysets, err := l.Client.ListKeySets(ctx, &admin.ListKeySetsRequest{
```
nit: Should call it `keySets` to stay consistent

tjasko commented on internal/resource/key_test.go:
```diff
@@ -0,0 +1,246 @@
+				{
+					Type:     v1.ErrorType_ERROR_TYPE_FIELD,
+					Field:    "pem.public_key",
+					Messages: []string{"'1234' is not valid 'pem-encoded-public-key'"},
```
Something I just thought of, should we modify `isPEMEncodedCertificate()` to ensure it's not expired? We don't have to do it in this PR, but realized we're accepting expired certificates.

tjasko commented on internal/resource/key.go:
```diff
@@ -0,0 +1,133 @@
+				Title:       "one key format",
+				Description: "Keys must be defined either in JWK or PEM format",
+				OneOf: []*generator.Schema{
+					{
+						Required: []string{"jwk"},
+						Properties: map[string]*generator.Schema{
+							"pem": {Not: &generator.Schema{Description: "there's a JWK, don't set PEM"}},
+						},
+					},
+					{
+						Required: []string{"pem"},
+						Properties: map[string]*generator.Schema{
+							"jwk": {Not: &generator.Schema{Description: "there's a PEM, don't set JWK"}},
+						},
+					},
```
Can we get this unit tested?

javierguerragiraldez replied:
it's there: https://github.com/Kong/koko/blob/0257e4ea283585f7b7a816f23e92fda81b359bed/internal/resource/key_test.go#L193-L212

tjasko replied:
But that doesn't assert the errors called out in this comment?

javierguerragiraldez commented on internal/grpc/proto/kong/admin/model/v1/key.proto:
```diff
@@ -0,0 +1,60 @@
+  int32 updated_at = 3;
+  repeated string tags = 4;
+
+  KeySet set = 5;
```
i followed the name already merged in Koko

javierguerragiraldez commented on internal/resource/key.go:
```diff
@@ -0,0 +1,133 @@
+}
+
+func (k Key) Indexes() []model.Index {
+	res := []model.Index{
```
i don't think so.   that sounds a bit information leaking, and not sure if it's even actually desirable.

tjasko replied:
Yeah I've seen it done either way too. I could go either way, mostly this just has concerns about the UX of the front-end... as it may be confusing to see duplicate keys where only one of them is used.

tjasko commented on internal/resource/key.go:
```diff
@@ -0,0 +1,133 @@
+	}
+	defaultID(&k.Key.Id)
+	if k.Key.Jwk != nil {
+		defaultID(&k.Key.Jwk.Kid)
```
Got it. Ah, missed that `defaultID()` no-ops when it's not empty, thanks!

javierguerragiraldez commented:
- made JWK field a JSON string, as Kong expects.
- added integration tests (which requires Kong/go-kong#238, so compilation fails)
- added JWK parseability validation (not full key validation)

GGabriele commented on internal/server/admin/keyset_service.go:
```diff
@@ -0,0 +1,144 @@
+	ctx context.Context,
+	req *v1.GetKeySetRequest,
+) (*v1.GetKeySetResponse, error) {
+	logger := s.logger(ctx)
```
Can we validate the presence of the ID for consistency with the other entities?

```
        if req.Id == "" {
		return nil, s.err(ctx, util.ErrClient{Message: "required ID is missing"})
	}
```

javierguerragiraldez replied:
done

GGabriele replied:
You will need to fix some tests like [this](https://github.com/Kong/koko/blob/6888d9f698f7416d7bd18e36ef2f56fdcaff333f/internal/server/admin/key_test.go#L108) after the change

GGabriele commented on internal/server/admin/key_test.go:
```diff
@@ -0,0 +1,285 @@
+	body.ContainsKey("kid")
+}
+
+func TestKeyCreate(t *testing.T) {
```
Can you please add a test creating a Key which references a KeySet?

javierguerragiraldez replied:
those are in the `TestKeyAndKeyset` function

GGabriele replied:
Ops, you're right

GGabriele reviewed (changes requested):
Just added a couple of minor asks. Looks great otherwise!
